extensions/ql-vscode/src/variant-analysis/variant-analysis-monitor.ts

```diff
@@ -121,6 +125,7 @@ export class VariantAnalysisMonitor extends DisposableObject {
 
       variantAnalysis = mapUpdatedVariantAnalysis(
         variantAnalysis,
+        this.getVariantAnalysisStatus(variantAnalysis.id),
```

Comment:
Could you explain why we call `getVariantAnalysisStatus(variantAnalysis.id)` to fetch the status from the variant analysis manager instead of using the `variantAnalysis.status` value we already have?

Reply:
Because the `variantAnalysis.status` we already have is one we got from the API so it won't necessarily have the "true" status (which is "canceling"). I should probably at least add a code comment to explain this.

Reply:
Hmm, I agree this is necessary for now, but I think it's not ideal. I'd like to look at changing it in a followup tech-debt issue.

In the code here, `variantAnalysisSummary` comes from the API, but the `variantAnalysis` variable we have here is of the internal type and has gone through mapping / processing. The problem is that it's a variable we keep within this `_monitorVariantAnalysis` function so it doesn't see any changes that come from outside the monitor. Previously this was fine because the monitor was the only thing that changed the variant analysis state and then it just fed data back to the manager. But now changes can come from other places and the monitor doesn't see these changes and tries to overwrite them.

Adding in `getVariantAnalysisStatus` works right now, but only because the `status` field is the only bit that's changing. It could cause problems if any other fields change in other places.

We could make it more general and fetch the whole variant analysis from the manager. We could also rewrite the monitor so it can track any changes by listening to the `onVariantAnalysisStatusUpdated` event from the manager.

Reply:
Yeah agree it's an area that could do with improvement. IMO it goes hand in hand with https://github.com/github/vscode-codeql/pull/3405#discussion_r1503890621.

Reply:
I've added a comment. I would like to make some incremental improvements around this and the mapping functions in separate PRs, if that's okay.